tensorflow/contrib/mpi_collectives/mpi_allreduce_test.py

```diff
@@ -0,0 +1,132 @@
+from __future__ import print_function
+
+import os
+import numpy as np
+import tensorflow as tf
+import tensorflow.contrib.mpi_collectives as mpi
+from tensorflow.python.platform import test
+
+
+average_allreduce = False
+max_wrong_count = -1
+
+
+class AllreduceTest(test.TestCase):
+  def dumpFailure(self, my_rank, out_loc_red, my_correct, out_all_red,
+                  our_correct):
+    # Find reduced/allreduced indices that are wrong and print all the
+    # values from output, slices, reduced, allreduced, so we can debug
+    # which is incorrect:
+    wrong_count = 0
+    red_dims = out_loc_red.shape
+    assert(len(red_dims) == 2)
+    for i in range(red_dims[0]):
+      for j in range(red_dims[1]):
+        suffix = ""
+        if out_loc_red[i][j] != my_correct[i][j] or \
+           out_all_red[i][j] != our_correct[i][j]:
+          suffix = "WRONG"
+          wrong_count += 1
+        print("{}\t{}\t{}\t{}\t{}\t{}"
+              .format(my_rank, i, j, out_loc_red[i][j],
+                      out_all_red[i][j], suffix), flush=True)
+        if max_wrong_count > 0 and wrong_count >= max_wrong_count:
+          return
+
+  def test_mpi_allreduce(self):
+    # Get MPI rank
+    my_rank = int(os.environ['PMI_RANK'])
+    num_ranks = int(os.environ['PMI_SIZE'])
+
+    stages = 13
+    batch_size = 1331
+    hidden_size = batch_size
+    out_size = batch_size
+
+    # Input placeholder (batch_size x hidden) - init to 1s
+    inputs = tf.placeholder(tf.float32, shape=(batch_size, hidden_size),
```

Comment:
We have only used internal RNN models to verify correctness of the allgather on IndexedSlices rather than unit tests. I'll create a test if you prefer.

Reply:
Up to you, but at least one small test seems appropriate.

Reply:
Added an Allgather test that runs correctly on our systems